### PR TITLE
Split epee/string_tools.h and extracted boost_lexical_cast

### PR DESCRIPTION
**Pull request by mj-xmr**
The epee's string_tools.h had 2 main problems: inline functions and a dependency on boost_lexical_cast, which is heavy enough. Everything that could have been un-inlined, is now in a .cpp file, and the 2 templates depending on boost_lexical_cast are now extracted. It turned out, that not all dependees needed both headers at the same time.
After this has been done, I have figured out, that the boost_lexical cast is still heavily used through other headers and thus I decided to get rid of these instances in one go.

Current dep graph of the string_tools:
now

Previous:
before

These are the dependees, which were interested in the lexical cast itself:
image
Still somehow heavy due to templated nature of them, but there will come time to fix this too.
Unfortunately the dependee graph of the string_tools wasn't generated.

Active work time: 4.8h

Speed improvements:
| Prevous   |      Current      |
|----------|-------------|
|Compilation (798 times): |  Compilation (807 times): |
| Parsing (frontend):         1728.2 s |     Parsing (frontend):         1725.6 s   | 
| Codegen & opts (backend):   1739.0 s | Codegen & opts (backend):   1744.8  |
| 80353 ms: boost/lexical_cast.hpp (included 179 times, avg 448 ms), included via: rctOps.cpp.o  (1121 ms) | Completely disappeared from the list | 

As you can see, there's no visible improvement in the total time for an unknown reason to me. Perhaps due to a larger number of TUs. But the supposidely 80s long (in total) compilation of  boost/lexical_cast.hpp has disappeared from the report:

[cba-result.txt](https://github.com/monero-project/monero/files/5536543/cba-result.txt)

### CODE REVIEW

**Comment by mj-xmr** [comment]
I'm not done yet. I will still try to extract the hex.h and mlocker.h from the string_tools.h, as they pull a lot of other dependencies.

**Comment by mj-xmr** [comment]
Resolved merge conflict.

File: contrib/epee/src/net_parse_helpers.cpp [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+#undef MONERO_DEFAULT_LOG_CATEGORY
+#define MONERO_DEFAULT_LOG_CATEGORY "net"
+
+namespace epee
+{
+namespace net_utils
+{
+
+  bool parse_uri_query(const std::string& query, std::list<std::pair<std::string, std::string> >& params)
```
Comment (vtnerd) [inline contrib/epee/src/net_parse_helpers.cpp:41]: I would drop this from the header and declare static or anonymous namespace. It is only called below, and can easily be changed later.

File: contrib/epee/src/string_tools.cpp [hunk 2]
```diff
@@ -68,6 +91,207 @@ namespace string_tools
...
+	 bool trim_right(std::string& str)
+	{
+
+		for(std::string::reverse_iterator it = str.rbegin(); it!= str.rend() && isspace(static_cast<unsigned char>(*it));)
+			str.erase( --((it++).base()));
+
+		return true;
+	}
+	//----------------------------------------------------------------------------
+	 std::string& trim(std::string& str)
```
Comment (vtnerd) [inline contrib/epee/src/string_tools.cpp:211]: I would leave this in the header.

File: contrib/epee/include/string_tools.h [hunk 3]
```diff
@@ -64,212 +45,26 @@ namespace epee
 {
 namespace string_tools
 {
-  //----------------------------------------------------------------------------
-  inline std::string buff_to_hex_nodelimer(const std::string& src)
```
[the remark this answers is not in the document]
Comment (mj-xmr) [reply contrib/epee/include/string_tools.h]: Done.

File: contrib/epee/include/misc_os_dependent.h [hunk 4]
```diff
@@ -47,7 +47,7 @@
 #endif
 
 #include <iostream>
-#include <boost/lexical_cast.hpp>
+#include <time.h>
```
[the remark this answers is not in the document]
Comment (mj-xmr) [reply contrib/epee/include/misc_os_dependent.h]: Done.

File: contrib/epee/include/misc_os_dependent.h [hunk 5]
```diff
@@ -115,15 +115,7 @@ namespace misc_utils
...
-
-	inline std::string get_thread_string_id()
-	{
-#if defined(_WIN32)
-		return boost::lexical_cast<std::string>(GetCurrentThreadId());
-#elif defined(__GNUC__)  
-		return boost::lexical_cast<std::string>(pthread_self());
-#endif
-	}
+	std::string get_thread_string_id();
```
[the remark this answers is not in the document]
Comment (mj-xmr) [reply contrib/epee/include/misc_os_dependent.h:118]: I added your comment with a TODO to the .cpp part.

File: contrib/epee/include/net/smtp.h [hunk 6]
```diff
@@ -33,118 +33,29 @@
 #include <ostream>
 #include <string>
 #include <boost/asio.hpp>
```
[the remark this answers is not in the document]
Comment (mj-xmr) [reply contrib/epee/include/net/smtp.h]: Done.

File: contrib/epee/include/string_tools.h [hunk 7]
```diff
@@ -24,33 +24,14 @@
...
-#include <boost/utility/string_ref.hpp>
-#include "misc_log_ex.h"
-#include "storages/parserse_base_utils.h"
 #include "hex.h"
-#include "memwipe.h"
 #include "mlocker.h"
-#include "span.h"
-#include "warnings.h"
 
+#include <string>
```
[the remark this answers is not in the document]
Comment (mj-xmr) [reply contrib/epee/include/string_tools.h]: Done.

File: contrib/epee/include/net/http_base.h [hunk 8]
```diff
@@ -66,44 +65,12 @@ namespace net_utils
...
-				if(!string_tools::compare_no_case(param_name, it->first))
-					break;
-
-			if(it==fields.end())
-				return std::string();
+		std::string get_value_from_fields_list(const std::string& param_name, const net_utils::http::fields_list& fields);
 
-			return it->second;
-		}
+        std::string get_value_from_uri_line(const std::string& param_name, const std::string& uri);
```
[the remark this answers is not in the document]
Comment (mj-xmr) [reply contrib/epee/include/net/http_base.h]: Done.

File: contrib/epee/src/tiny_ini.cpp [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+		std::string expr_str = std::string() + "^("+ param_name +") *=(.*?)$";
+		const boost::regex match_ini_entry( expr_str, boost::regex::icase | boost::regex::normal); 
+		boost::smatch result;	
+		if(!boost::regex_search(ini_entry, result, match_ini_entry, boost::match_default))
+			return false;
+		res = result[2];
+		string_tools::trim(res);
+		return true;
+	}
+    std::string get_param_value(const std::string& param_name, const std::string& ini_entry)
```
[the remark this answers is not in the document]
Comment (mj-xmr) [reply contrib/epee/src/tiny_ini.cpp]: Done.

File: contrib/epee/src/http_base.cpp [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+        }
+        return std::string();
+    }
+
+    void add_field(std::string& out, const boost::string_ref name, const boost::string_ref value)
+    {
+        out.append(name.data(), name.size()).append(": ");
+        out.append(value.data(), value.size()).append("\r\n");
+    }
+    void add_field(std::string& out, const std::pair<std::string, std::string>& field)
```
[the remark this answers is not in the document]
Comment (mj-xmr) [reply contrib/epee/src/http_base.cpp]: Done.

File: contrib/epee/src/net_parse_helpers.cpp (hunk 1 above)
Comment (mj-xmr) [reply contrib/epee/src/net_parse_helpers.cpp:41]: I just dropped it from the header and left it in the .cpp file untouched.

File: contrib/epee/src/string_tools.cpp (hunk 2 above)
Comment (mj-xmr) [reply contrib/epee/src/string_tools.cpp]: Done.

**Comment by mj-xmr** [comment]
Resolved conflict at `contrib/epee/src/CMakeLists.txt`

**Comment by mj-xmr** [comment]
Rebased to resolve test problems.

File: contrib/epee/include/net/http_base.h [hunk 11]
```diff
@@ -136,22 +110,22 @@ namespace net_utils
...
 		};
 
-    struct uri_content
-    {
-      std::string m_path;
-      std::string m_query;
-      std::string m_fragment;
-      std::list<std::pair<std::string, std::string> > m_query_params;
-    };
+	struct uri_content
```
Comment (vtnerd) [inline contrib/epee/include/net/http_base.h:113]: Revert whitespace only changes in this file :/ kind of a rule around here.

File: contrib/epee/src/net_parse_helpers.cpp (hunk 1 above)
Comment (vtnerd) [reply contrib/epee/src/net_parse_helpers.cpp:41]: There's no need to export the function symbol. The `static` keyword here does the same thing without having to mess with spacing.

**Review by vtnerd — COMMENTED** [review]
The comments are trivial (and not strictly necessary) - but I'll probably have to re-scan for changes if something is pushed.

File: contrib/epee/src/net_parse_helpers.cpp (hunk 1 above)
Comment (mj-xmr) [reply contrib/epee/src/net_parse_helpers.cpp]: There will be need for exporting this once we get serious with unit testing, but for now I added the `static` kw

**Comment by mj-xmr** [comment]
> The comments are trivial (and not strictly necessary) - but I'll probably have to re-scan for changes if something is pushed.

Rhanis for getting back to it. I will rescan them myself soon as well.

File: contrib/epee/include/net/http_base.h (hunk 11 above)
Comment (mj-xmr) [reply contrib/epee/include/net/http_base.h]: Done.

**Review by vtnerd — APPROVED** [review]

**Review by selsta — APPROVED** [review]

**Pull request merged**